src/shared/Core/Trace2.cs

```diff
@@ -0,0 +1,278 @@
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1026:DefaultParametersShouldNotBeUsed")]
+    void WriteVersion(
+        string gcmVersion,
+        string eventFormatVersion = "3",
+        [System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
+        [System.Runtime.CompilerServices.CallerLineNumber] int lineNumber = 0);
+
+    /// <summary>
+    /// Writes argv and time information associated with application start to
+    /// trace writer.
+    /// </summary>
+    /// <param name="argv">Exe name and arguments.</param>
+    /// <param name="timeInMilliseconds">Elapsed time in milliseconds since GCM started.</param>
+    /// <param name="filePath">Path of the file this method is called from.</param>
+    /// <param name="lineNumber">Line number of file this method is called from.</param>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1026:DefaultParametersShouldNotBeUsed")]
+    void WriteStart(
+        IList<string> argv,
+        double timeInMilliseconds,
+        [System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
+        [System.Runtime.CompilerServices.CallerLineNumber] int lineNumber = 0);
+    
+    /// <summary>
+    /// Writes exit code and execution time to trace writer.
+    /// </summary>
+    /// <param name="code">Application exit code.</param>
+    /// <param name="timeInMilliseconds">Elapsed time in milliseconds since GCM started.</param>
+    /// <param name="filePath">Path of the file this method is called from.</param>
+    /// <param name="lineNumber">Line number of file this method is called from.</param>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1026:DefaultParametersShouldNotBeUsed")]
+    void WriteExit(
+        int code,
+        double timeInMilliseconds,
+        [System.Runtime.CompilerServices.CallerFilePath]
+        string filePath = "",
+        [System.Runtime.CompilerServices.CallerLineNumber]
+        int lineNumber = 0);
+}
+
+public class Trace2: DisposableObject, ITrace2
+{
+    private const string ParentSidVariable = "GIT_TRACE2_PARENT_SID";
+    private const string CurrentSidVariable = "GCM_TRACE2_PARENT_SID";
+    private readonly ICommandContext _context;
+    private bool _isDisposed;
+    private CollectorWriter _writer;
+
+    public Trace2(ICommandContext context)
+    {
+        _context = context;
+    }
+
+    public string GetSid()
+    {
+        var parentSid = "";
+        if (_context.Environment.Variables.ContainsKey(ParentSidVariable))
+        {
+            parentSid = _context.Environment.Variables.FirstOrDefault(name => 
+                name.Key.Equals(ParentSidVariable, StringComparison.OrdinalIgnoreCase)).Value + "/";
+        }
+
+        var sidSuffix = Guid.NewGuid();
+        var currentSid = $"{parentSid}{sidSuffix}";
+        
+        _context.Environment.SetEnvironmentVariable(CurrentSidVariable, currentSid);
```

Comment:
Oh, I see, you're writing out a new GCM_...SID rather that overwriting the inherited GIT_...SID envvar.
That's why it doesn't keep getting longer with each call.

Instead, we should overwrite the GIT_...SID envvar so that when GCM invoked `git config`, the child will inherit the modified SID.  This will put GCM in the trace-graph between the `git fetch` and the `git config`, for example.

Reply:
I wonder if we should just go all in on `GIT_TRACE2` and not introduce `GCM_` specific variables.